src/groupeddataframe/splitapplycombine.jl

```diff
@@ -962,14 +1056,51 @@ function _combine(f::AbstractVector{<:Pair},
                 throw(ArgumentError("a single value or vector result is required when passing " *
                                     "multiple functions (got $(typeof(res)))"))
             end
+            if !(firstres isa AbstractVector) && isnothing(idx_agg)
+                idx_agg = Vector{Int}(undef, length(gd))
+                fillfirst!(nothing, idx_agg, 1:length(gd.groups), gd)
+            end
             idx, outcols, _ = _combine_with_first(wrap(firstres), fun, gd, incols,
-                                                  Val(firstmulticol))
+                                                  Val(firstmulticol),
+                                                  firstres isa AbstractVector ? nothing : idx_agg)
+            @assert length(outcols) == 1
             res[i] = idx, outcols[1]
         end
     end
-    # TODO: avoid recomputing idx for each pair when !all(isagg, f)
-    idx = res[1][1]
+    idx_loc = findfirst(x -> x[1] !== idx_agg, res)
+    if isnothing(idx_loc)
+        @assert !isnothing(idx_agg)
+        idx = idx_agg
+    else
+        idx = res[idx_loc][1]
+        agg2idx_map = nothing
+        for i in 1:length(res)
+            if res[i][1] != idx
+                if res[i][1] === idx_agg
+                    # we perform pseudo broadcasting here
+                    # keep -1 as a sentinel for errors
+                    if isnothing(agg2idx_map)
+                        agg2idx_map = fill(-1, length(idx))
+                        aggj = 1
+                        for j in 1:length(idx)
+                            while idx_agg[aggj] != idx[j]
+                                aggj += 1
+                            end
+                            agg2idx_map[j] = aggj
+                        end
+                    end
+                    res[i] = idx, res[i][2][agg2idx_map]
+                else
+                    if idx != res[i][1]
+                        throw(ArgumentError("all functions must return values of the same length"))
+                    end
+                end
+            end
+        end
+    end
     outcols = map(x -> x[2], res)
+    # this check is redundant given we check idx above now
+    # but it is safer to double check and it is cheap
     if !all(x -> length(x) == length(outcols[1]), outcols)
         throw(ArgumentError("all functions must return values of the same length"))
     end
```

Comment:
Then it's no longer a user error:
```suggestion
    @assert all(x -> length(x) == length(outcols[1]), outcols)
```

Reply:
Right :smile: